### DISCUSSION TITLE
Am I hitting limitations with generics, or am I misusing the framework?

### DISCUSSION THREAD

**Discussion by PGijsbers** [pydantic/pydantic#5750] [Question] [answered]
I have read the [docs on data conversion](https://docs.pydantic.dev/latest/usage/models/#data-conversion), but I think I have hit some limitations when working with generics. Specifically, I want to keep information around (and specifically enforce) that a generic reference is to specific (sub)classes:

```python
from __future__ import annotations

from abc import ABC
from typing import TypeVar, Generic

from pydantic import BaseModel


class BaseObject(BaseModel, ABC):
    id_: str


T = TypeVar("T", bound=BaseObject)


class ObjectReference(BaseModel, Generic[T]):
    id_: str

    @classmethod
    def to(cls, object_: T) -> ObjectReference[T]:
        return cls(id_=object_.id_)


class Address(BaseObject):
    street: str
    number: int


class Job(BaseObject):
    role: str
    company: str


class Person(BaseObject):
    name: str
    job: ObjectReference[Job]


if __name__ == "__main__":
    job = Job(id_="engineer", role="engineer", company="FizzBuzz Co.")
    alice = Person(id_="alice", name="alice", job=ObjectReference.to(job))
    bob = Person(id_="bob", name="bob", job=job)  # How is this converted to a reference?

    address = Address(id_="home", street="Foobarsquare", number=42)
    catlyn = Person(id_="catlyn", name="catlyn", job=address)  # Wrong type of reference.
    dylan = Person(id_="dylan", name="dylan", job=ObjectReference.to(address))  # Wrong type of reference.

```

I assume `bob`'s job reference is created through something like `job.dict(include={"id_"})`, but now the exact type of reference is lost, yet Pydantic doesn't complain. A similar thing likely happens `catlyn` and `dylan`'s jobs, where the latter even explicitly is of the wrong type. I also can't get `mypy` to pick up on these errors.
This leads me to the question, am I simply hitting some limitations of the tools/design decisions, or did I set this up incorrectly?

Thanks a lot!

**Comment by dmontagu** [comment]
It doesn't look like there is any runtime information that could be used to determine what the generic parameter of the `ObjectReference` instance is; this may be part of why pydantic can't help. 

When I try to run your code example though I get an error on the `bob = ...` line. This is with pydantic v2, which I assumed is what you were using because you have `class ObjectReference(BaseModel, Generic[T]):`, which only works in v2. Prior to that, you should use `from pydantic.generics import GenericModel` if attempting to create a generic pydantic model. (Though I might  be misunderstanding what you are trying to accomplish.)

Because I can't tell exactly what you are trying to do, I'm not confident in my assessment of whether what you are doing makes sense, but I _suspect_ there is some subtle misuse of generic types here. I do think whatever you are trying to do may be possible to express with the type system, but it will have some limitations.

If you can give a specific example of code you want to succeed and/or cause errors, that might help us better understand how to help.

**Comment by PGijsbers** [comment]
Thanks for taking the time to look at my question! 🙏 

The setting is that these object will be used in some JSON-LD graph, in which case the objects don't contain each other directly, but merely include a reference to their ID. E.g.:
```json
{
  "@context": "...",
  "@graph": [
    {
      "@id": "person-alice",
      "@type": "Person",
      "name": "Alice",
      "job": {
        "@id": "fizzbuzz-engineer",
      }
    },
    {
      "@id": "fizzbuzz-engineer",
      "@type": "Job",
      "role": "engineer",
      "company": "FizzBuzz Co.",
    }
  ]
```
In some settings there might be properties which have lists of references, so just storing the `id` directly as a `str` won't do.
I want to keep some type information around to make sure the right types are referenced.

It is possible to achieve this serialization behavior by storing a direct reference and then just changing serialisation:
```python
from pydantic import field_serializer

# BaseObject from code in OP

class Person(BaseObject):
    name: str
    job: Job

    @field_serializer('job')
    def serialize_job(self, job: Job, _info):
        return job.model_dump_json(include={"id_"})
```
But I would prefer an option which can be deserialized. The code in the OP works for that case if I would do away with the `Generic` aspect of it, and simply have an `ObjectReference` (as opposed to `ObjectReference[T]`). However, I thought I could keep information around on _which_ types should be "referenced", both for code readability and hopefully type validation (runtime and/or static).

I have tried `pydantic==1.10.7` with `GenericModel` which leads to the behavior in OP (interestingly, `BaseModel` also seems to exhibit the same behavior - sorry for not catching that and throwing you off). 

When using `pydantic==2.0a4` I interestingly encounter a different error than you do (MacOS 13.1, Python 3.11.3):
```
Traceback (most recent call last):
  File "~/repositories/tmp/limits/example.py", line 41, in <module>
    alice = Person(id_="alice", name="alice", job=ObjectReference.to(job))
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/repositories/tmp/limits/venv/lib/python3.11/site-packages/pydantic/main.py", line 246, in __init__
    __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
pydantic_core._pydantic_core.ValidationError: 1 validation error for Person
job
  Input should be a valid dictionary [type=dict_type, input_value=ObjectReference(id_='engineer'), input_type=ObjectReference]
    For further information visit https://errors.pydantic.dev/0.30.0/v/dict_type
```

The following change makes sure `alice` gets initialised and an error occurs when initialising `bob`:
```diff
class Person(BaseObject):
    name: str
-    job: ObjectReference[Job]
+    job: ObjectReference
```
but this is not the solution I am looking for as this loses type information again.

**Reply by dmontagu** [reply]
This error is coming from the fact that `ObjectReference[Job]` is a proper subclass of `ObjectReference`, and so when you have a field of type `ObjectReference[Job]`, it's not letting you pass something of type `ObjectReference` in. (We need to update that error message that says should be a valid dictionary.)

I think we can make this work better, but let me think about if there's a different way to structure the code and/or set up a custom type..

**Reply by dmontagu** [reply, marked as answer]
One thing that might make it work for you would be to change `ObjectReference.to` to return `return cls[type(object_)](id_=object_.id_)`

**Reply by PGijsbers** [reply]
Thanks! As far as I can tell, this works as expected on `2.0a4` (for runtime type checking) :)
I tried `cls[T](id_object_.id_)` before, so I was close. After thinking about it I understand why that wouldn't work (`T` is not the specific type, but just any `T`-type).

It still doesn't seem to work under `pydantic==1.10.7` though (with `class ObjectReference(GenericModel, Generic[T]):`). Since there is a V2 answer, I marked it as such (I can use it soon enough). I am still curious about the V1 situation, so I would be happy with some help/pointers, but I appreciate that you might have better things to do :-)

edit: and when I say it doesn't work in v1, I mean that it happily instantiates all the objects despite their types not matching.

**Reply by dmontagu** [reply]
If you can show me the behavior you are getting in v2 that you aren't getting in v1 (ideally via a code sample I can "just run", which it seems you have handy), I can try to explain why it misbehaves in v1 and/or how to get it working in v1.

That said, we improved a lot of things related to generics in v2 so I won't be very surprised if it's a consequence of a difficult-to-resolve bug in v1 that probably won't get fixed.
